### ISSUE TITLE
Ancient block number issue

### ISSUE THREAD

**Issue by vintidui** [openethereum/openethereum#206]
- **OpenEthereum version (>=3.1.0)**: 3.1.1-rc.1-stable-71bc626-20201221
- **Operating system**: Linux
- **Installation**: one-line installer / built from source
- **Fully synchronized**: yes
- **Network**: ethereum
- **Restarted**: yes

I was depending on version 3.1.0 and kept running into similar issues to https://github.com/openethereum/openethereum/issues/69. Since that issue corrupted the DB each time, I saved a copy of the database. Each time the DB corrupted, I just replaced the corrupted DB with my backup and re-synced from that point in order to avoid a full sync from scratch. I don't think this is causing my problem, but thought that it might be useful information.

The issue I keep running into is this error: https://github.com/openethereum/openethereum/blob/main/ethcore/src/client/client.rs#L535

The good news is that it I don't think it corrupts the database because I'm able to restart after the error kills the process. It looks like it was introduced in `3.1.1-rc.1`. I've noticed that the ancient block header number does not seem to be predictable as it seems to "randomly" increase _and_ decrease, at different times. So I could see the possibility of the ancient block number jumping too far ahead somehow.

Here's some output from when a run went wrong and I had to restart:

```
2020-12-28 15:01:40   Ancient:#11504157 AB:#11509459   27/50 peers   5 MiB chain 199 KiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:01:46  Imported #11544508 0x5271…374c (120 txs, 12.42 Mgas, 1602 ms, 34.65 KiB)
2020-12-28 15:01:49  Imported #11544509 0x3c1b…c82f (167 txs, 12.45 Mgas, 1095 ms, 41.91 KiB)
2020-12-28 15:02:03  Imported #11544510 0x88b3…c598 (225 txs, 12.44 Mgas, 995 ms, 48.68 KiB)
2020-12-28 15:02:05  Imported #11544510 0xa80b…f2de (211 txs, 12.47 Mgas, 908 ms, 46.71 KiB)
2020-12-28 15:02:11   Ancient:#11504157 AB:#11511530   26/50 peers   4 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:02:13  Imported #11544511 0x31b6…d163 (151 txs, 12.45 Mgas, 1252 ms, 46.90 KiB)
2020-12-28 15:02:44  Imported #11544512 0xefeb…2509 (199 txs, 12.46 Mgas, 714 ms, 48.32 KiB)
2020-12-28 15:02:50  Imported #11544513 0x5eb9…5011 (217 txs, 12.46 Mgas, 994 ms, 47.85 KiB)
2020-12-28 15:02:50   Ancient:#11504157 AB:#11516499   27/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:03:12  Imported #11544515 0x0139…e49e (131 txs, 12.45 Mgas, 1167 ms, 35.99 KiB)
2020-12-28 15:03:25   Ancient:#11504157 AB:#11519858   26/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:03:31  Imported #11544516 0x90f7…1ea0 (143 txs, 12.45 Mgas, 937 ms, 63.04 KiB)
2020-12-28 15:03:44  Imported #11544517 0x3fa3…e6dc (194 txs, 12.47 Mgas, 1271 ms, 50.13 KiB)
2020-12-28 15:03:50  Imported #11544518 0x0816…13bc (140 txs, 12.47 Mgas, 1117 ms, 38.05 KiB)
2020-12-28 15:03:56   Ancient:#11504157 AB:#11521156   26/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:03:58  Imported #11544519 0x48aa…26c0 (156 txs, 12.50 Mgas, 1220 ms, 40.62 KiB)
2020-12-28 15:04:03  Imported #11544520 0x3be5…fd3e (312 txs, 12.49 Mgas, 1280 ms, 49.41 KiB)
2020-12-28 15:04:30   Ancient:#11504157 AB:#11523634   25/25 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:04:34  Imported #11544521 0x014e…4d68 (207 txs, 12.51 Mgas, 4011 ms, 45.17 KiB)
2020-12-28 15:04:38  Imported #11544522 0x35e6…b127 (174 txs, 12.51 Mgas, 2428 ms, 50.45 KiB)
2020-12-28 15:04:44  Imported #11544523 0x9472…70d3 (150 txs, 12.52 Mgas, 1115 ms, 45.32 KiB)
2020-12-28 15:05:00   Ancient:#11504157 AB:#11525654   26/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:05:02  Imported #11544524 0x9dc9…eacc (276 txs, 12.52 Mgas, 1047 ms, 50.35 KiB)
2020-12-28 15:05:20  Imported #11544525 0x7866…ab91 (206 txs, 12.50 Mgas, 1189 ms, 47.31 KiB)
2020-12-28 15:05:35  Imported #11544526 0xec22…6704 (188 txs, 12.49 Mgas, 1099 ms, 54.95 KiB)
2020-12-28 15:05:40   Ancient:#11504157 AB:#11528783   26/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:05:54  Imported #11544527 0x5784…e404 (232 txs, 12.50 Mgas, 1152 ms, 59.88 KiB)
2020-12-28 15:06:12  Imported #11544528 0xb5ec…85d7 (182 txs, 12.49 Mgas, 1177 ms, 40.83 KiB)
2020-12-28 15:06:16  Imported #11544529 0xab07…ccb5 (138 txs, 12.47 Mgas, 926 ms, 50.08 KiB)
2020-12-28 15:06:18   Ancient:#11504157 AB:#11531916   27/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:06:36  Imported #11544530 0x261f…c5ac (164 txs, 12.49 Mgas, 960 ms, 48.67 KiB)
2020-12-28 15:06:45  Imported #11544531 0xdd89…9efc (148 txs, 12.49 Mgas, 1288 ms, 38.08 KiB)
2020-12-28 15:06:50  Imported #11544532 0x4191…3958 (170 txs, 12.35 Mgas, 1304 ms, 43.86 KiB)
2020-12-28 15:06:54   Ancient:#11504157 AB:#11533755   27/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:06:59  Imported #11544533 0x088c…6dd0 (136 txs, 12.50 Mgas, 1255 ms, 36.21 KiB)
2020-12-28 15:07:06  Imported #11544534 0xad24…7241 (148 txs, 12.49 Mgas, 1054 ms, 45.91 KiB)
2020-12-28 15:07:08  Imported #11544535 0x0a42…3fd9 (177 txs, 12.49 Mgas, 1341 ms, 42.89 KiB)
2020-12-28 15:07:20  Imported #11544536 0x0143…d8ec (181 txs, 12.49 Mgas, 1450 ms, 43.34 KiB)
2020-12-28 15:07:29  Imported #11544537 0xaf01…10a0 (151 txs, 12.50 Mgas, 2204 ms, 34.11 KiB)
2020-12-28 15:07:30   Ancient:#11504157 AB:#11504281   27/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:07:36  Imported #11544538 0x2cd6…86fb (166 txs, 12.49 Mgas, 1295 ms, 40.56 KiB)
2020-12-28 15:07:41  Imported #11544539 0xaf73…16c5 (137 txs, 12.48 Mgas, 1311 ms, 37.99 KiB)
2020-12-28 15:07:50  Imported #11544540 0x777c…1153 (148 txs, 12.49 Mgas, 1411 ms, 40.52 KiB)
2020-12-28 15:07:54  Imported #11544541 0x878e…78c3 (120 txs, 12.46 Mgas, 2737 ms, 32.45 KiB)
2020-12-28 15:08:01   Ancient:#11504157 AB:#11505324   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:08:13  Imported #11544542 0x8d73…f0fa (167 txs, 12.46 Mgas, 1166 ms, 54.33 KiB)
2020-12-28 15:08:30  Imported #11544543 0x2f76…7499 (171 txs, 12.47 Mgas, 1085 ms, 41.89 KiB)
2020-12-28 15:08:33  Imported #11544544 0xcb39…41e9 (149 txs, 12.48 Mgas, 1302 ms, 38.32 KiB)
2020-12-28 15:08:37   Ancient:#11504157 AB:#11507212   27/50 peers   7 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:08:45  Imported #11544545 0xf77c…48da (161 txs, 12.50 Mgas, 1160 ms, 42.44 KiB)
2020-12-28 15:09:10   Ancient:#11504157 AB:#11510072   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:09:10  Imported #11544546 0xf878…9d5d (198 txs, 12.47 Mgas, 1122 ms, 49.95 KiB)
2020-12-28 15:09:38  Imported #11544547 0x8ea9…7741 (225 txs, 12.48 Mgas, 1106 ms, 52.70 KiB)
2020-12-28 15:09:46   Ancient:#11504157 AB:#11513205   27/50 peers   7 MiB chain 56 KiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:09:48  Reorg to #11544549 0x337b…59d4 (0x9f10…7471 #11544547 0x8ea9…7741 0x3f52…7da1)
2020-12-28 15:09:49  Imported #11544550 0x2de4…e93a (144 txs, 12.19 Mgas, 1141 ms, 38.16 KiB)
2020-12-28 15:09:58  Imported #11544551 0x2b88…c3a0 (165 txs, 12.47 Mgas, 1176 ms, 42.41 KiB)
2020-12-28 15:10:05  Imported #11544552 0x5736…fadb (9 txs, 12.47 Mgas, 958 ms, 24.99 KiB)
2020-12-28 15:10:14  Imported #11544553 0x8135…3b44 (166 txs, 12.48 Mgas, 1350 ms, 51.79 KiB)
2020-12-28 15:10:21   Ancient:#11504157 AB:#11514142   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:10:22  Imported #11544554 0x8599…8039 (131 txs, 12.47 Mgas, 1787 ms, 35.56 KiB)
2020-12-28 15:10:31  Imported #11544555 0x6354…edc8 (163 txs, 12.48 Mgas, 1490 ms, 37.51 KiB)
2020-12-28 15:10:55   Ancient:#11504157 AB:#11515965   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:11:18  Imported #11544556 0x861f…8474 (231 txs, 12.47 Mgas, 1549 ms, 46.53 KiB)
2020-12-28 15:11:24  Imported #11544557 0x5b15…9b28 (159 txs, 12.49 Mgas, 1354 ms, 45.97 KiB)
2020-12-28 15:11:28   Ancient:#11504157 AB:#11517030   26/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:11:30  Imported #11544558 0x8e18…ea51 (170 txs, 12.47 Mgas, 1450 ms, 56.30 KiB)
2020-12-28 15:11:34  Imported #11544559 0x5c26…4174 (63 txs, 12.48 Mgas, 640 ms, 41.72 KiB)
2020-12-28 15:11:46  Imported #11544560 0x2c33…9225 (201 txs, 12.48 Mgas, 1368 ms, 47.46 KiB)
2020-12-28 15:12:00   Ancient:#11504157 AB:#11519221   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:12:03  Imported #11544561 0x683e…f27e (209 txs, 12.49 Mgas, 1136 ms, 37.04 KiB)
2020-12-28 15:12:11  Imported #11544562 0xa1c5…a1ff (141 txs, 12.50 Mgas, 1207 ms, 36.87 KiB)
2020-12-28 15:12:17  Imported #11544563 0x7275…a596 (117 txs, 12.48 Mgas, 1592 ms, 34.73 KiB)
2020-12-28 15:12:25  Imported #11544564 0x6e62…2c81 (276 txs, 12.49 Mgas, 1143 ms, 45.69 KiB)
2020-12-28 15:12:33  Imported #11544565 0x8d5e…ea74 (165 txs, 12.47 Mgas, 1615 ms, 54.50 KiB)
2020-12-28 15:12:36   Ancient:#11504157 AB:#11520313   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:12:46  Imported #11544566 0xa63a…94ff (216 txs, 12.48 Mgas, 1233 ms, 54.27 KiB)
2020-12-28 15:12:50  Imported #11544567 0x3592…3736 (187 txs, 12.49 Mgas, 1407 ms, 43.36 KiB)
2020-12-28 15:13:10   Ancient:#11504157 AB:#11521923   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:13:27  Imported #11544568 0xaf8b…9c75 (249 txs, 12.48 Mgas, 1281 ms, 57.95 KiB)
2020-12-28 15:13:40   Ancient:#11504157 AB:#11524193   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:13:50  Imported #11544569 0xa456…a219 (203 txs, 12.48 Mgas, 1035 ms, 47.28 KiB)
2020-12-28 15:14:10   Ancient:#11504157 AB:#11527151   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:14:38  Imported #11544570 0x88ab…46b4 (236 txs, 12.50 Mgas, 992 ms, 49.16 KiB)
2020-12-28 15:14:45  Imported #11544571 0x737c…a0ef (172 txs, 12.49 Mgas, 1252 ms, 43.85 KiB)
2020-12-28 15:14:47   Ancient:#11504157 AB:#11532201   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:14:53  Imported #11544572 0x9d33…a14e (162 txs, 12.48 Mgas, 1143 ms, 44.97 KiB)
2020-12-28 15:15:20   Ancient:#11504157 AB:#11535058   27/50 peers   6 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:15:24  Imported #11544573 0xb952…c7f1 (280 txs, 12.47 Mgas, 907 ms, 45.61 KiB)
2020-12-28 15:15:46  Imported #11544574 0xd764…3a1e (182 txs, 12.49 Mgas, 1081 ms, 46.64 KiB)
2020-12-28 15:15:53   Ancient:#11504157 AB:#11538523   27/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:16:00  Imported #11544575 0x39d7…d238 (196 txs, 12.50 Mgas, 1229 ms, 51.72 KiB)
2020-12-28 15:16:25   Ancient:#11504157 AB:#11541780   27/50 peers   5 MiB chain 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 15:16:39  Panic occured, see stderr for details


====================

stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: start_thread
  14: clone


Thread 'IO Worker #0' panicked at 'Ancient block number is higher then best block number', ethcore/src/client/client.rs:535

This is a bug. Please report it at:

    https://github.com/openethereum/openethereum/issues/new


2020-12-28 15:16:39  Finishing work, please wait...
2020-12-28 15:16:39  Imported #11544576 0x285d…70f7 (196 txs, 12.47 Mgas, 993 ms, 42.85 KiB)
Error: 1
vintidui@ethtop:~$ openethereum
2020-12-28 17:57:08  Starting OpenEthereum/v3.1.1-rc.1-stable-71bc626-20201221/x86_64-linux-gnu/rustc1.48.0
2020-12-28 17:57:08  Keys path /home/vintidui/.local/share/openethereum/keys/ethereum
2020-12-28 17:57:08  DB path /home/vintidui/.local/share/openethereum/chains/ethereum/db/906a34e69aec8c0d
2020-12-28 17:57:08  State DB configuration: fast
2020-12-28 17:57:08  Operating mode: active
2020-12-28 17:57:08  Configured for Ethereum using Ethash engine
2020-12-28 17:57:09  Downloading old blocks from 0xdde4cbf1e6a9080495d0fcca6055f08319cbe3064235d55fbdeb44fc6f792752 (#11504157) till Some(0x743865760032bf972447e3c58a2e224e120f702580f665e737a4f5eb3f748276) (#Some(11460001))
2020-12-28 17:57:09  Listening for new connections on 127.0.0.1:8546.
2020-12-28 17:57:09  Updated conversion rate to Ξ1 = US$725.40 (6564522.5 wei/gas)
2020-12-28 17:57:14  Public node URL: <Manually removed this info as I'm unsure if it's safe to include>
2020-12-28 17:57:19  Syncing #11544576 0x285d…70f7     0.00 blk/s    0.0 tx/s    0.0 Mgas/s     43+   76 Qed Ancient:#11504157 LI:#11544797   13/25 peers   117 KiB chain 18 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 17:57:24  Syncing #11544579 0x571d…d23f     0.60 blk/s  109.6 tx/s    7.5 Mgas/s      0+  250 Qed Ancient:#11504157 LI:#11544830   16/25 peers   378 KiB chain 42 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 17:57:29  Syncing #11544582 0x9999…0e67     0.60 blk/s  109.6 tx/s    7.5 Mgas/s      0+  247 Qed Ancient:#11504157 LI:#11544830   18/25 peers   583 KiB chain 41 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 17:57:34  Syncing #11544585 0xb7d3…2e40     0.60 blk/s   88.6 tx/s    7.5 Mgas/s      0+  244 Qed Ancient:#11504157 LI:#11544830   16/25 peers   766 KiB chain 41 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 17:57:39  Syncing #11544587 0x5195…e067     0.40 blk/s   75.4 tx/s    4.9 Mgas/s      0+  242 Qed Ancient:#11504157 LI:#11544830   20/25 peers   906 KiB chain 40 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 17:57:44  Syncing #11544591 0xc889…158f     0.80 blk/s  134.2 tx/s   10.0 Mgas/s      0+  238 Qed Ancient:#11504157 LI:#11544931   21/25 peers   907 KiB chain 40 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 17:57:49  Syncing #11544594 0x94d0…410e     0.60 blk/s   94.6 tx/s    7.5 Mgas/s      0+  336 Qed Ancient:#11504157 LI:#11544931   21/25 peers   1 MiB chain 55 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2020-12-28 17:57:54  Syncing #11544597 0x1216…b86e     0.60 blk/s   84.2 tx/s    7.5 Mgas/s      0+  333 Qed Ancient:#11504157 LI:#11544931   25/25 peers   1 MiB chain 55 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
```

Please let me know what I can do to help debug. I recently joined the discord channel with my same username. Thanks in advance!

**Comment by rakita** [comment]
It is same problem as #69, but I put that condition to panic, basically, ancient block gets higher than the best block and that can already potentially mean that db has some invalid states.
You will need to resync again.

**Comment by vintidui** [comment]
Got it. Thanks for your speedy response. I cleared the DB with `openethereum db kill --chain=ethereum` and started over. Interestingly enough, snapshots stalled at 300/5802 the first time. I restarted and now I'm cruising at 874/6356. I've had issues with inconsistent snapshot hops before. 5802 vs 6356 makes a big difference on where I end up in the chain which, of course, significantly affects the duration of the total sync. Any idea why total snapshots appear to be random? And even if snapshotting happens at all? Regardless, I'll report back on any interesting findings.
